File: nn.py
from abc import ABC, abstractmethod
import pickle
from typing import Callable, Any

from tensor import Tensor, RingTensor


class Partial:
    """Wrappper for a function with fixed arguments and nice repr."""
    def __init__(self, fn: Callable, *args, **kwargs):
        self.fn = fn
        self.args = args
        self.kwargs = kwargs

    def __call__(self, x):
        return self.fn(x, *self.args, **self.kwargs)

    def __repr__(self):
        args_repr = ', '.join([f"{arg!r}" for arg in self.args] + [f"{k}={v!r}" for k, v in self.kwargs.items()])
        return f"{self.fn.__name__}({args_repr})"


class Model(ABC):
    weights: list[Tensor]
    nparams: int

    @abstractmethod
    def __call__(self, x: Tensor) -> Tensor:
        raise NotImplementedError

    @abstractmethod
    def save(self, path: str):
        raise NotImplementedError

    @abstractmethod
    def load(self, path: str):
        raise NotImplementedError


class Module(Model):
    _weights: list[Tensor] = []

    @property
    def weights(self) -> list[Tensor]:
        all_weights = self._weights or []
        for m in self.__dict__.values():
            if isinstance(m, Module):
                all_weights += m.weights
        return all_weights

    @property
    def nparams(self) -> int:
        return sum(w.size for w in self.weights)

    def __call__(self, x: Tensor) -> Tensor:
        return self.forward(x)

    def forward(self, x: Tensor) -> Tensor:
        raise NotImplementedError

    def save(self, path: str):
        with open(path, 'wb') as f:
            pickle.dump(self.weights, f)

    def load(self, path: str):
        with open(path, 'rb') as f:
            for old, new in zip(self.weights, pickle.load(f)):
                old.set_to(new)
        return self


class RingFF(Module):
    def __init__(self, input_size: int, output_size: int):
        self._weights = [RingTensor.rand((input_size, output_size), requires_grad=True)]
        # We don't need any bias, because it would be mathematically equivalent to just shifting the weights of the following layer by the corresponding amount.

    def forward(self, x: RingTensor) -> RingTensor:
        return (x.unsqueeze(-1) - self._weights[0]).cos().mean(axis=-2)
        # return (x.unsqueeze(-1) - self._weights[0]).poly_sigmoid(1.2, 4).mean(axis=-2)

    def __repr__(self):
        return f"RingFF(input_size={self._weights[0].shape[0]}, output_size={self._weights[0].shape[1]})"


class RingConv(Module):
    def __init__(self, input_size: int, output_size: int, window_size: int = 3, padding: int = 1, stride: int = 1):
        self.window_size = window_size
        self.padding = padding
        self.stride = stride
        self._weights = [RingTensor.rand((input_size, window_size, window_size, output_size), requires_grad=True)]

    def forward(self, x: RingTensor) -> RingTensor:
        return (x.sliding_window_2d(self.window_size, self.padding, self.stride).unsqueeze(-1) - self._weights[0]).cos().mean(axis=(-4,-3,-2))
        # return (x.sliding_window_2d(self.window_size, self.padding, self.stride).unsqueeze(-1) - self._weights[0]).poly_sigmoid(1.2, 4).mean(axis=(-4,-3,-2))

    def __repr__(self):
        return f"RingConv(input_size={self._weights[0].shape[0]}, output_size={self._weights[0].shape[3]}, window_size={self.window_size}, padding={self.padding}, stride={self.stride})"


class Sequential(Module):
    def __init__(self, modules: list[Module | Callable[[Any],Tensor]]):
        self.modules = modules

    @property
    def weights(self):
        return [w for m in self.modules if isinstance(m, Module) for w in m.weights]

    def forward(self, x: Tensor) -> Tensor:
        for m in self.modules:
            x = m(x)
        return x

    def __repr__(self):
        steps = []
        for m in self.modules:
            if isinstance(m, Module):
                steps.append(m.__repr__())
            elif isinstance(m, Partial):
                steps.append(m.__repr__())
            elif hasattr(m, '__name__') and m.__name__ == '<lambda>' and hasattr(m, '__code__'):
                steps.append(f"lambda {', '.join(m.__code__.co_varnames)}: ...")
            else:
                steps.append(str(m))
        return "\n-> ".join(steps)


class Input(Model):
    # placeholder methods for type annotations
    def ff(self, output_size: int) -> "Input": ...
    def conv(self, output_size: int, window_size: int = 3, padding: int = 1, stride: int = 1) -> "Input": ...
    def apply(self, fn: Callable[[RingTensor], Tensor]) -> "Input": ...
    def flatten(self, *axes: int) -> "Input": ...
    def save(self, path: str): ...
    def load(self, path: str) -> "Sequential": ...
    # also Tensor methods are available

    def __init__(self, shape: tuple[int, ...]):
        self._input_shape = shape
        self._shape = shape
        self._network = Sequential([])

    def __call__(self, x: Tensor) -> Tensor:
        return object.__getattribute__(self, "_network")(x)

    def __getattribute__(self, name: str) -> Any:
        shape = object.__getattribute__(self, "_shape")
        network = object.__getattribute__(self, "_network")

        def add_fn(fn: Callable[[RingTensor], Tensor]) -> "Input":
            output_shape = fn(RingTensor.rand(shape)).shape
            network.modules.append(fn)
            object.__setattr__(self, "_shape", output_shape)
            return self

        match name:
            case "ff":
                return lambda output_size: add_fn(RingFF(shape[-1], output_size))
            case "conv":
                return lambda output_size, window_size=3, padding=1, stride=1: add_fn(RingConv(shape[-1], output_size, window_size, padding, stride))
            case "apply":
                return lambda fn: add_fn(fn)
            case "weights" | "nparams" | "save" | "load":
                return getattr(network, name)
            case _:
                if hasattr(Tensor, name):
                    # add tensor operation to the network
                    return lambda *args, **kwargs: add_fn(Partial(getattr(Tensor, name), *args, **kwargs))
                raise AttributeError(f"Input has no attribute {name}")

    def __repr__(self):
        '''representation including the input and output shapes'''
        input_shape = object.__getattribute__(self, "_input_shape")
        output_shape = object.__getattribute__(self, "_shape")
        network = object.__getattribute__(self, "_network")
        return f"Input(shape={input_shape})\n-> {network!r}\n-> Output(shape={output_shape})"

File: nn_visualizer/nn/src/nn.ts
import { RingTensor, RealTensor, Tensor } from './tensor.js';
import type { AnyTensor } from './types.js';

export abstract class Model {
  abstract get weights(): AnyTensor[];
  abstract get nparams(): number;
  abstract forward(x: AnyTensor): AnyTensor;
  call(x: AnyTensor): AnyTensor { return this.forward(x); }
}

export abstract class Module extends Model {
  protected _weights: AnyTensor[] = [];

  get weights(): AnyTensor[] {
    const own = this._weights ?? [];
    const nested: AnyTensor[] = [];
    for (const v of Object.values(this as any)) {
      if (v instanceof Module) nested.push(...v.weights);
    }
    return [...own, ...nested];
  }
  get nparams(): number { return this.weights.reduce((a, w) => a + w.size, 0); }
  forward(_x: AnyTensor): AnyTensor { throw new Error('Not implemented'); }
}

export class RingFF extends Module {
  constructor(inputSize: number, outputSize: number) {
    super();
    this._weights = [RingTensor.rand([inputSize, outputSize])];
  }
  forward(x: RingTensor): RingTensor {
    // (x.unsqueeze(-1) - W).cos().mean(axis=-2)
    const expanded = x.unsqueeze(-1) as RingTensor;
    const diff = expanded.sub(this._weights[0] as RingTensor) as RingTensor;
    const act = diff.cos();
    return act.mean(-2) as RingTensor;
  }
  toString(): string {
    const w = this._weights[0];
    return `RingFF(input_size=${w.shape[0]}, output_size=${w.shape[1]})`;
  }
}

export class RingConv extends Module {
  private window: number;
  private padding: number;
  private stride: number;
  constructor(inputChannels: number, outputChannels: number, windowSize = 3, padding = 1, stride = 1) {
    super();
    this.window = windowSize; this.padding = padding; this.stride = stride;
    this._weights = [RingTensor.rand([inputChannels, windowSize, windowSize, outputChannels])];
  }
  forward(x: RingTensor): RingTensor {
    // (x.sliding_window_2d(ws,pad,stride).unsqueeze(-1) - W).cos().mean(axis=(-4,-3,-2))
    const sw = x.sliding_window_2d(this.window, this.padding, this.stride) as RingTensor;
    const expanded = sw.unsqueeze(-1) as RingTensor;
    const diff = expanded.sub(this._weights[0] as RingTensor) as RingTensor;
    const act = diff.cos();
    // mean over last 3 window dims: (-4,-3,-2) - use array to mean over multiple axes at once
    // After unsqueeze(-1), shape is (B, H', W', C, ws, ws, 1)
    // After subtraction with weights (C, ws, ws, O), shape is (B, H', W', C, ws, ws, O)
    // So -4,-3,-2 refer to dimensions C, ws, ws (indices 3, 4, 5 in rank-7 tensor)
    const rank = act.shape.length;
    const axes = [rank - 4, rank - 3, rank - 2]; // Map negative indices to positive
    return act.mean(axes, false) as RingTensor;
  }
  toString(): string {
    const w = this._weights[0];
    return `RingConv(input_size=${w.shape[0]}, output_size=${w.shape[3]}, window_size=${this.window}, padding=${this.padding}, stride=${this.stride})`;
  }
}

export class Sequential extends Module {
  constructor(readonly modules: (Module | ((x: any) => any))[]) { super(); }
  get weights(): AnyTensor[] { return this.modules.flatMap(m => m instanceof Module ? m.weights : []); }
  forward(x: any, onLayerOutput?: (layerIdx: number, name: string, output: any) => void): any {
    let value = x;
    for (let i = 0; i < this.modules.length; i++) {
      const m = this.modules[i];
      value = m instanceof Module ? m.forward(value) : m(value);
      if (onLayerOutput) {
        const name = m instanceof Module ? m.toString() : String(m);
        onLayerOutput(i, name, value);
      }
    }
    return value;
  }
  toString(): string { return this.modules.map(m => (m instanceof Module ? m.toString() : String(m))).join('\n-> '); }
}

export class Partial<F extends (x: any, ...args: any[]) => any> {
  constructor(public fn: F, public args: any[] = [], public kwargs: Record<string, any> = {}) {}
  call = (x: any) => this.fn(x, ...this.args);
  toString(): string {
    const argsRepr = [...this.args.map(a => JSON.stringify(a)), ...Object.entries(this.kwargs).map(([k, v]) => `${k}=${JSON.stringify(v)}`)].join(', ');
    return `${this.fn.name}(${argsRepr})`;
  }
}

export class Input extends Model {
  private _inputShape: number[];
  private _shape: number[];
  private _network: Sequential;

  constructor(shape: number[]) {
    super();
    this._inputShape = [...shape];
    this._shape = [...shape];
    this._network = new Sequential([]);
  }
  get weights(): AnyTensor[] { return this._network.weights; }
  get nparams(): number { return this._network.nparams; }
  forward(x: any, onLayerOutput?: (layerIdx: number, name: string, output: any) => void): any {
    return this._network.forward(x, onLayerOutput);
  }

  private addFn<T extends AnyTensor>(fn: (x: any) => T): this {
    // Infer output shape by running on a dummy tensor
    const dummy = RingTensor.rand(this._shape);
    const out = fn(dummy);
    this._shape = [...out.shape];
    this._network.modules.push(fn as any);
    return this;
  }

  ff(outputSize: number): this {
    const inputSize = this._shape[this._shape.length - 1];
    const module = new RingFF(inputSize, outputSize);
    this._network.modules.push(module);
    const dummy = RingTensor.rand(this._shape);
    const out = module.forward(dummy);
    this._shape = [...out.shape];
    return this;
  }

  conv(outputSize: number, windowSize = 3, padding = 1, stride = 1): this {
    const inputChannels = this._shape[this._shape.length - 1];
    const module = new RingConv(inputChannels, outputSize, windowSize, padding, stride);
    this._network.modules.push(module);
    const dummy = RingTensor.rand(this._shape);
    const out = module.forward(dummy);
    this._shape = [...out.shape];
    return this;
  }

  apply(fn: (x: RingTensor) => AnyTensor): this { return this.addFn(fn); }
  flatten(...axes: number[]): this { return this.addFn((x: RingTensor) => x.flatten(...axes)); }

  toString(): string {
    return `Input(shape=${JSON.stringify(this._inputShape)})\n-> ${this._network.toString()}\n-> Output(shape=${JSON.stringify(this._shape)})`;
  }
}

export { RingTensor, RealTensor, Tensor };
export type { AnyTensor } from './types.js';


